src/Bicep.Core/Semantics/SymbolicReferenceCollector.cs

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Collections.Generic;
+using System.Linq;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.Semantics;
+
+internal class SymbolicReferenceCollector : AstVisitor
+{
+    private readonly HashSet<DeclaredSymbol> symbolsReferenced = new();
+    private readonly IBinder binder;
+
+    private SymbolicReferenceCollector(IBinder binder)
+    {
+        this.binder = binder;
+    }
+
+    internal static IEnumerable<DeclaredSymbol> CollectSymbolsReferenced(IBinder binder, DeclaredSymbol symbol)
+    {
+        SymbolicReferenceCollector collector = new(binder);
+        symbol.DeclaringSyntax.Accept(collector);
+        return collector.symbolsReferenced;
+    }
+
+    internal static IEnumerable<DeclaredSymbol> CollectSymbolsReferencedRecursive(IBinder binder, DeclaredSymbol symbol)
```

Comment:
I'm surprised this functionality doesn't already exist in the codebase!

Reply:
We do something very similar in `CyclicCheckVisitor`, but I wasn't able to reuse that code path since there are some special cases where we permit syntactic cycles.